_posts/2023-06-07-airflow-upgrade-to-2-5-3.md

```diff
@@ -0,0 +1,129 @@
+- applying migrations took ~2 hours on our weak dev RDS cluster and we thought it will be faster on production... boy oh boy we were wrong...
+- our plain Airflow `User` role lost access to the custom _Self-Service Backfill UI_. Solved this by setting `AIRFLOW__WEBSERVER__UPDATE_FAB_PERMS` to `False` because otherwise it drops custom permissions each time Airflow Web Server restarts
+
+#### Ship it!
+
+Finally, everything was tested and fixed. The time has come to upgrade Airflow on production! All DAGs paused, ECS tasks stopped. Production database snapshot created.
+
+###### Database migration
+We started applying migrations on production database. In order to do this we prepared another ECS task which runs `airfow db upgrade` command. We successfully deployed it via Terraform and start to wait... and wait... and wait a little bit more...
+
+After ~5 hours we started to think why it takes so long... We realized that:
+- our production Airflow database RDS cluster only twice as big as our development
+- we have giant database with data for 2017 year and so on (backfills mostly). Considering we had a `sentinel` DAG which checks Scheduler is alive for each 5 minutes (sends a gauge metric to the Datadog), and our main DAG with ~1k tasks in it running daily, you can imagine how big `task_instance` and `dag_run` tables were...
+- ???Missed index in the database
+
+
+What we decided to do:
+- stop the migration
+- give our production RDS cluster more resources - 32Gb RAM and 16 CPUs comparing to 8Gb RAM and 4 CPUs that we had before
+- delete all data below year 2021
+- rerun migration ECS task again
+
+That helped - migration completed in ~30 minutes!
+
+###### Airflow - your turn!
+
+Deploy ECS tasks via Terraform with updated Airflow image didn't cause any issues. 
+
+After that we've merged all PR's with updated DAGs and related services. Checked that there are no DAG parsing errors on production.
+
+Started all DAGs and start to observe... and it started to fail...
+- [Braking changes](https://github.com/apache/airflow/pull/26452/files#diff-55326294fdc9ce88aee820373ed658972dff4067517a9c4f59819efbbf3e3b85R27) in the SlackWebHook. Basically they refactored the hook and switched to the Slack SDK instead of using plain HTTP. This caused `SlackRequestError: Invalid URL detected: ***` for all our Slack hook usages. In order to fix it we had to went through all Slack Airflow Connections, copy `webhook_token` extra param, change the connection type to Slack Incoming Webhook and put copied value to the `Webhook Token` field
+- ???Bug in Databricks SQL operator they are creating new connection for each query. This breaks our templated queries because in most of them first query is `use some_database` and next is an actual processing query. But when it creates new connection for each query in a file it simply forgets about use statement and fails second query because can't find tables without database
```

Comment:
I don't like this point because I'd like to create an upstream PR with a fix first and then just put here a link to the upstream PR instead of describing it on fingers...